src/parser/shared/metrics/activeTime.ts

```diff
@@ -0,0 +1,70 @@
+import { AnyEvent, EventType } from 'parser/core/Events';
+import metric, { Info } from 'parser/core/metric';
+import { ClosedTimePeriod, duration, intersection, union } from 'parser/core/timePeriods';
+
+// TODO possible extensions:
+//   * channel and GCD time periods marked with spell, allowing additional filtering
+
+/** Time periods the player was channeling during the encounter */
+export const channelTimePeriods = metric((events: AnyEvent[], info: Info): ClosedTimePeriod[] => {
+  /*
+   * Method:
+   * All completed channels can be calculated by just using the endchannel event,
+   * which links back to the matching beginchannel. Cancelled channels must be ignored because
+   * there's no way of knowing when they were cancelled. The one special case is a channel that is
+   * started but the encounter ends before it finishes.
+   */
+  let uncoveredUncancelledBeginChannelTime: number | undefined = undefined;
+  const results: ClosedTimePeriod[] = events.reduce((acc: ClosedTimePeriod[], event) => {
+    if (event.type === EventType.BeginChannel && !event.isCancelled) {
+      uncoveredUncancelledBeginChannelTime = event.timestamp;
+    } else if (event.type === EventType.EndChannel) {
+      acc.push({ start: event.beginChannel.timestamp, end: event.timestamp });
+      uncoveredUncancelledBeginChannelTime = undefined;
+    }
+    return acc;
+  }, []);
+  // handle special case of channel in progress when encounter ends
```

Comment:
Ooh this is a tough one. For active time I imagine you'd want to calculate this as downtime, seeing as it's practically a cancelled cast that didn't land. It might help to mark this special case with an `end: null`?

Reply:
I don't feel I need a special time period because it's not this metric's job to judge that, it just wants to know when the player was 'active' at channeling - so capping at the encounter end makes sense to me and also makes the downstream processing a lot more straightforward.

I did consider counting this channel as 'cancelled', all I'd have to do is just remove this special case handling. It didn't feel 'fair' to the player, though.

Reply:
I think it's fair since players can play around it by casting an instant cast or auto attacking. Casting a 3 sec cast while the boss is dying in 1.5 is the wrong decision.